src/system/SystemPacketBuffer.cpp

```diff
@@ -605,14 +605,34 @@ PacketBufferHandle PacketBufferHandle::PopHead()
     return PacketBufferHandle(head);
 }
 
-PacketBufferHandle PacketBufferHandle::CloneData(uint16_t aAdditionalSize, uint16_t aReservedSize)
+PacketBufferHandle PacketBufferHandle::CloneData()
 {
-    if (!mBuffer->Next().IsNull())
+    PacketBufferHandle cloneHead;
+
+    for (PacketBuffer * original = mBuffer; original != nullptr; original = static_cast<PacketBuffer *>(original->next))
     {
-        // We do not clone an entire chain.
-        return PacketBufferHandle();
+        uint16_t originalDataSize     = original->MaxDataLength();
+        uint16_t originalReservedSize = original->ReservedSize();
+        PacketBufferHandle clone      = PacketBufferHandle::New(originalDataSize, originalReservedSize);
+        if (clone.IsNull())
+        {
+            return PacketBufferHandle();
```

Comment:
I think this is a memory leak now if the buffer chain has 3 or more segments and New fails for the 3rd or any subsequent link.

This error path needs to call `Free` on `clone before returning.

Reply:
`PacketBufferHandle`'s destructor will free things as needed here, no?

Reply:
`cloneHead` is a `PacketBufferHandle`, so it'll be freed as it goes out of scope.

Reply:
Ah, I see.  Please disregard.